# Apriori Association Rule Mining

Apriori association rules is a machine learning method used to analyze relations between variables such as basket analysis (items that are purchased together).

To demonstrate how to use apriori association rules, I will use grocery store data from Kaggle: https://www.kaggle.com/ekrembayar/apriori-association-rules-grocery-store/data

In this dataset, we have rows of grocery items that are purchased together.

## Import libraries

In [141]:
import pandas as pd
import os
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

## Data

In [142]:
path = 'C:/Users/Katia/Documents/Machine learning'
os.chdir(path)

df = pd.read_csv('grocery_data_whole_foods.csv', sep = ",")

In [143]:
df.head()

1                    2               3  \
0      citrus fruit  semi-finished bread       margarine   
1    tropical fruit               yogurt          coffee   
2        whole milk                  NaN             NaN   
3         pip fruit               yogurt    cream cheese   
4  other vegetables           whole milk  condensed milk   

                          4    5    6    7    8    9   10  ...   23   24   25  \
0               ready soups  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1                       NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
2                       NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
3              meat spreads  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
4  long life bakery product  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   

    26   27   28   29   30   31   32  
0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 32 columns]

## Data Coversion

In [144]:
# Convert the dataframe into sparce data, necessary for using the apriori function

sparce_df = pd.get_dummies(df)

In [145]:
sparce_df.head()

1_Instant food products  1_UHT-milk  1_abrasive cleaner  \
0                        0           0                   0   
1                        0           0                   0   
2                        0           0                   0   
3                        0           0                   0   
4                        0           0                   0   

   1_artif. sweetener  1_baby cosmetics  1_bags  1_baking powder  \
0                   0                 0       0                0   
1                   0                 0       0                0   
2                   0                 0       0                0   
3                   0                 0       0                0   
4                   0                 0       0                0   

   1_bathroom cleaner  1_beef  1_berries  ...  28_chocolate  \
0                   0       0          0  ...             0   
1                   0       0          0  ...             0   
2                   0       0          0  ...             0   
3                   0       0          0  ...             0   
4                   0       0          0  ...             0   

   28_hygiene articles  28_napkins  28_sugar  29_cooking chocolate  \
0                    0           0         0                     0   
1                    0           0         0                     0   
2                    0           0         0                     0   
3                    0           0         0                     0   
4                    0           0         0                     0   

   29_house keeping products  29_soups  30_skin care  31_hygiene articles  \
0                          0         0             0                    0   
1                          0         0             0                    0   
2                          0         0             0                    0   
3                          0         0             0                    0   
4                          0         0             0                    0   

   32_candles  
0           0  
1           0  
2           0  
3           0  
4           0  

[5 rows x 2201 columns]

## Support Calculation

Support is a metric of how frequently an item appears in the dataset. For example, if an item is purchased 50 times out of a total of 10,000 transactions, support will be equal to 0.005.

In [146]:
# Get support table (how popular an item is) with length value (number of items purchased together)

df = sparce_df

frequent_itemsets = apriori(df, min_support=0.002, use_colnames=True)
# you can be more conservative regaring the minimum support value, but it might exclude combinations of 3 or more items
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets

support                                           itemsets  length
0    0.007117                                       (1_UHT-milk)       1
1    0.030910                                           (1_beef)       1
2    0.012303                                        (1_berries)       1
3    0.008134                                      (1_beverages)       1
4    0.018099                                   (1_bottled beer)       1
..        ...                                                ...     ...
798  0.002339       (5_yogurt, 3_other vegetables, 4_whole milk)       3
799  0.004169  (5_whole milk, 4_other vegetables, 3_root vege...       3
800  0.002034                   (4_curd, 5_yogurt, 3_whole milk)       3
801  0.002135       (5_whole milk, 6_butter, 4_other vegetables)       3
802  0.002339  (4_root vegetables, 5_other vegetables, 6_whol...       3

[803 rows x 3 columns]

In [148]:
# Filter the length of item combinations (length = 1 is not useful for basket analysis). 
# Here I will demonstrate with length = 3

frequent_itemsets[frequent_itemsets['length'] >= 3]

support                                           itemsets  length
783  0.003152    (3_pip fruit, 2_tropical fruit, 1_citrus fruit)       3
784  0.002440       (1_other vegetables, 2_whole milk, 3_yogurt)       3
785  0.003254  (1_root vegetables, 3_whole milk, 2_other vege...       3
786  0.002034      (3_other vegetables, 4_whole milk, 1_sausage)       3
787  0.002440      (5_whole milk, 4_other vegetables, 1_sausage)       3
788  0.002135  (2_other vegetables, 3_whole milk, 1_tropical ...       3
789  0.002237       (4_butter, 3_whole milk, 2_other vegetables)       3
790  0.003254       (3_whole milk, 2_other vegetables, 4_yogurt)       3
791  0.002135    (4_whole milk, 3_other vegetables, 2_pip fruit)       3
792  0.005186  (4_whole milk, 3_other vegetables, 2_root vege...       3
793  0.002034  (2_tropical fruit, 4_whole milk, 3_other veget...       3
794  0.002135  (3_pip fruit, 2_tropical fruit, 4_other vegeta...       3
795  0.002237  (2_tropical fruit, 4_other vegetables, 3_root ...       3
796  0.002135  (5_whole milk, 2_tropical fruit, 4_other veget...       3
797  0.002847       (3_other vegetables, 4_whole milk, 5_butter)       3
798  0.002339       (5_yogurt, 3_other vegetables, 4_whole milk)       3
799  0.004169  (5_whole milk, 4_other vegetables, 3_root vege...       3
800  0.002034                   (4_curd, 5_yogurt, 3_whole milk)       3
801  0.002135       (5_whole milk, 6_butter, 4_other vegetables)       3
802  0.002339  (4_root vegetables, 5_other vegetables, 6_whol...       3

The most frequent 3-item combination includes whole milk, other vegetables, and root vegetables.

# Association Rules

## Confidence

Confidence is a matric of how frequent the rule has been found to be true. Highest value is 1. 

In [149]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.7) 
# I am being conservative here with the threshold for simplification purposes
rules

antecedents           consequents  \
0                          (2_liquor)      (1_bottled beer)   
1                         (2_sausage)       (1_frankfurter)   
2                      (6_whole milk)  (5_other vegetables)   
3                          (6_butter)        (5_whole milk)   
4                      (7_whole milk)  (6_other vegetables)   
5                          (7_butter)        (6_whole milk)   
6       (3_pip fruit, 1_citrus fruit)    (2_tropical fruit)   
7      (1_other vegetables, 3_yogurt)        (2_whole milk)   
8   (1_root vegetables, 3_whole milk)  (2_other vegetables)   
9      (4_butter, 2_other vegetables)        (3_whole milk)   
10  (2_root vegetables, 4_whole milk)  (3_other vegetables)   
11   (5_whole milk, 2_tropical fruit)  (4_other vegetables)   
12     (3_other vegetables, 5_butter)        (4_whole milk)   
13  (5_whole milk, 3_root vegetables)  (4_other vegetables)   
14           (5_whole milk, 6_butter)  (4_other vegetables)   
15     (6_butter, 4_other vegetables)        (5_whole milk)   
16  (4_root vegetables, 6_whole milk)  (5_other vegetables)   

    antecedent support  consequent support   support  confidence        lift  \
0             0.003152            0.018099  0.002237    0.709677   39.211671   
1             0.010066            0.058973  0.010066    1.000000   16.956897   
2             0.008846            0.012201  0.007016    0.793103   65.001437   
3             0.003660            0.015150  0.002949    0.805556   53.172073   
4             0.004982            0.006914  0.004067    0.816327  118.067227   
5             0.003050            0.008846  0.002339    0.766667   86.668582   
6             0.003152            0.036096  0.003152    1.000000   27.704225   
7             0.003355            0.066497  0.002440    0.727273   10.936892   
8             0.004169            0.055923  0.003254    0.780488   13.956541   
9             0.002339            0.051449  0.002237    0.956522   18.591682   
10            0.006202            0.042196  0.005186    0.836066   19.813747   
11            0.002745            0.025826  0.002135    0.777778   30.115923   
12            0.002847            0.032028  0.002847    1.000000   31.222222   
13            0.004982            0.025826  0.004169    0.836735   32.398763   
14            0.002949            0.025826  0.002135    0.724138   28.038963   
15            0.002237            0.015150  0.002135    0.954545   63.006406   
16            0.002745            0.012201  0.002339    0.851852   69.816358   

    leverage  conviction  
0   0.002180    3.382105  
1   0.009472         inf  
2   0.006908    4.774360  
3   0.002893    5.064943  
4   0.004033    5.406801  
5   0.002312    4.247803  
6   0.003038         inf  
7   0.002217    3.422844  
8   0.003021    4.300796  
9   0.002117   21.816675  
10  0.004924    5.842603  
11  0.002064    4.383782  
12  0.002756         inf  
13  0.004040    5.966815  
14  0.002059    3.531380  
15  0.002101   21.666701  
16  0.002305    6.667641

The item combination with the highest support and confidence is sausage and frankfurter, with 100% confidence.

## Lift

"This says how likely item Y is purchased when item X is purchased, while controlling for how popular item Y is." definition from https://www.kdnuggets.com/2016/04/association-rules-apriori-algorithm-tutorial.html
Lift > 1, item X is highly associated with Y. 
Lift < 1, item X is poorly associated with Y.
Lift = 1, item X and Y are not associated.

In [150]:
# Get lift scores

rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules

antecedents  \
0                      (2_citrus fruit)   
1                              (1_beef)   
2                              (1_beef)   
3                  (2_other vegetables)   
4                   (2_root vegetables)   
..                                  ...   
785   (4_root vegetables, 6_whole milk)   
786  (5_other vegetables, 6_whole milk)   
787                 (4_root vegetables)   
788                (5_other vegetables)   
789                      (6_whole milk)   

                                 consequents  antecedent support  \
0                                   (1_beef)            0.024403   
1                           (2_citrus fruit)            0.030910   
2                       (2_other vegetables)            0.030910   
3                                   (1_beef)            0.055923   
4                                   (1_beef)            0.038943   
..                                       ...                 ...   
785                     (5_other vegetables)            0.002745   
786                      (4_root vegetables)            0.007016   
787       (5_other vegetables, 6_whole milk)            0.010981   
788        (4_root vegetables, 6_whole milk)            0.012201   
789  (4_root vegetables, 5_other vegetables)            0.008846   

     consequent support   support  confidence       lift  leverage  conviction  
0              0.030910  0.004881    0.200000   6.470395  0.004126    1.211362  
1              0.024403  0.004881    0.157895   6.470395  0.004126    1.158522  
2              0.055923  0.003254    0.105263   1.882297  0.001525    1.055145  
3              0.030910  0.003254    0.058182   1.882297  0.001525    1.028957  
4              0.030910  0.005592    0.143603   4.645845  0.004389    1.131590  
..                  ...       ...         ...        ...       ...         ...  
785            0.012201  0.002339    0.851852  69.816358  0.002305    6.667641  
786            0.010981  0.002339    0.333333  30.354938  0.002262    1.483528  
787            0.007016  0.002339    0.212963  30.354938  0.002262    1.261674  
788            0.002745  0.002339    0.191667  69.816358  0.002305    1.233717  
789            0.004372  0.002339    0.264368  60.466453  0.002300    1.353432  

[790 rows x 9 columns]

Let's look at items 785 and 786. We can see that if root vegetables and whole milk are purchased, there is a good chance that other vegetables will also be purchase, as suggested both by the lift value and confidence value. However, even though these three items are highly associated, there is less confidence on the purchase of root vegetables if other vegetables and whole milk are purchased first, as suggested by the low confidence value of 786.

## Lift and Confidence

In [124]:
# Filter rules for optimal lift and confidence

rules[(rules['lift'] >= 5) & (rules['confidence']>= 0.5)]

antecedents  \
20                                (2_liquor)   
92                               (2_sausage)   
100                                 (3_pork)   
214                                  (2_ham)   
218                                 (2_meat)   
380                                (2_herbs)   
417                                 (3_beef)   
444                            (3_pip fruit)   
468                                 (3_curd)   
510                               (3_onions)   
513                           (4_whole milk)   
555                            (4_pip fruit)   
562                               (4_butter)   
564                                 (4_curd)   
598                               (4_onions)   
600                           (5_whole milk)   
604                               (6_butter)   
621                               (5_butter)   
642                     (5_other vegetables)   
643                           (6_whole milk)   
649                      (5_root vegetables)   
651                               (6_butter)   
660                           (7_whole milk)   
661                     (6_other vegetables)   
662                               (7_butter)   
671            (3_pip fruit, 1_citrus fruit)   
677           (1_other vegetables, 3_yogurt)   
682        (1_root vegetables, 3_whole milk)   
695                (5_whole milk, 1_sausage)   
707           (4_butter, 2_other vegetables)   
714           (2_other vegetables, 4_yogurt)   
719              (2_pip fruit, 4_whole milk)   
725        (2_root vegetables, 4_whole milk)   
737        (3_pip fruit, 4_other vegetables)   
748         (5_whole milk, 2_tropical fruit)   
755           (3_other vegetables, 5_butter)   
756                 (4_whole milk, 5_butter)   
760           (5_yogurt, 3_other vegetables)   
767        (5_whole milk, 3_root vegetables)   
772                       (4_curd, 5_yogurt)   
778                 (5_whole milk, 6_butter)   
780           (6_butter, 4_other vegetables)   
782                               (6_butter)   
784  (4_root vegetables, 5_other vegetables)   
785        (4_root vegetables, 6_whole milk)   

                            consequents  antecedent support  \
20                     (1_bottled beer)            0.003152   
92                      (1_frankfurter)            0.010066   
100                     (1_frankfurter)            0.004270   
214                         (1_sausage)            0.006507   
218                         (1_sausage)            0.007626   
380                (3_other vegetables)            0.004881   
417                            (2_pork)            0.004474   
444                  (2_tropical fruit)            0.014947   
468                      (2_whole milk)            0.010168   
510                (4_other vegetables)            0.007524   
513                (3_other vegetables)            0.032028   
555                  (3_tropical fruit)            0.007321   
562                      (3_whole milk)            0.009151   
564                      (3_whole milk)            0.008846   
598                (5_other vegetables)            0.003457   
600                (4_other vegetables)            0.015150   
604                (4_other vegetables)            0.003660   
621                      (4_whole milk)            0.006812   
642                      (6_whole milk)            0.012201   
643                (5_other vegetables)            0.008846   
649                (6_other vegetables)            0.005287   
651                      (5_whole milk)            0.003660   
660                (6_other vegetables)            0.004982   
661                      (7_whole milk)            0.006914   
662                      (6_whole milk)            0.003050   
671                  (2_tropical fruit)            0.003152   
677                      (2_whole milk)            0.003355   
682                (2_other vegetables)            0.004169 

Filtering rules can help prioritize products for upsale and cross-sale, without getting lost in super long lists.